src/test/java/net/openhft/chronicle/bytes/BytesReleaseInvariantNonPerformantMethodsTest.java

```diff
@@ -0,0 +1,175 @@
+        final AtomicReference<BytesInitialInfo> initialInfo = new AtomicReference<>();
+        return cartesianProductTest(BytesFactoryUtil::provideBytesObjects,
+                BytesReleaseInvariantNonPerformantMethodsTest::provideNonPerformantOperations,
+                (args, bytes, nc) -> {
+                    if (bytes.refCount() > 0) {
+                        if (isReadWrite(args)) {
+                            bytes.write(SILLY_NAME);
+                        }
+                        initialInfo.set(new BytesInitialInfo(bytes));
+                        bytes.releaseLast();
+                    }
+                    final String name = createCommand(args) + "->" + bytes(args).getClass().getSimpleName() + "." + nc.name();
+
+                    assertThrows(ClosedIllegalStateException.class, () -> nc.accept(bytes), name);
+
+                    // Unable to check actual size for released MappedBytes
+                    if (!MappedBytes.class.isInstance(bytes)) {
+                        // Todo: remove this check once the cursors are invalid for released resources
+                        final BytesInitialInfo info = new BytesInitialInfo(bytes);
+                        assertEquals(initialInfo.get(), info, name);
+                    }
+                }
+        );
+    }
+
+    /**
+     * Checks the bytes.toDebugString() works with released resources
+     */
+    @ParameterizedTest
+    @MethodSource("net.openhft.chronicle.bytes.BytesFactoryUtil#provideBytesObjects")
+    void toDebugString(final Bytes<?> bytes, final boolean readWrite) {
+        toDebug(bytes, readWrite, Bytes::toDebugString);
+    }
+
+    /**
+     * Checks the bytes.toDebugString(10) works with released resources
+     */
+    @ParameterizedTest
+    @MethodSource("net.openhft.chronicle.bytes.BytesFactoryUtil#provideBytesObjects")
+    void toDebugString10(final Bytes<?> bytes, final boolean readWrite) {
+        toDebug(bytes, readWrite, b -> b.toDebugString(10));
+    }
+
+    // Rather than throwing an exception, toDebug provides a String instead
+    private void toDebug(final Bytes<?> bytes, final boolean readWrite, Function<? super Bytes<?>, String> operation) {
+        if (readWrite) {
+            bytes.append(SILLY_NAME);
+        }
+        releaseAndAssertReleased(bytes);
+        final String actual = operation.apply(bytes);
+        assertEquals("<released>", actual);
+    }
+
+    private static Stream<NamedConsumer<Bytes<Object>>> provideNonPerformantOperations() {
+        final OutputStream os = new OutputStream() {
+            @Override
+            public void write(int b) throws IOException {
+                throw new UnsupportedEncodingException();
+            }
+        };
+        final BytesStore<?, ?> bs = BytesStore.from(SILLY_NAME);
+        final Bytes<?> bytes = Bytes.from(SILLY_NAME);
+        return Stream.of(
+                NamedConsumer.of(Bytes::toHexString, "toHexString()"),
```

Comment:
consider using static import for the NamedConsumer

Reply:
@RobAustin strongly disagree - lots of things have an "of" method - `NamedConsumer.of` reads nicely and is clear